File: mobile/android/app/src/main/java/com/findly/android/ui/designsystem/components/FindlyCodeDisplay.kt
package com.findly.android.ui.designsystem.components

import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.sp
import com.findly.android.ui.designsystem.FindlyTheme

/**
 * Review-round addition (A36, specs/010-app-shell-and-screen-ux.md §5.1: "The code, large and
 * copyable: `titleLarge`-class size, tabular figures, letter-spaced, in hyphenated display
 * form"). `titleLarge` alone covers the size; tabular figures and extra letter-spacing need a
 * value the six §4.1 type roles don't carry, and 003 §4.3 forbids a screen from reaching for a
 * literal `TextStyle.copy(letterSpacing = …)` itself. This component is where that value-level
 * derivation belongs instead — the same precedent [FindlyMapMarkerBubble]'s `selected` flag and
 * its private `MarkerNameStyle`/`NowPillStyle`/`StaleAgeStyle` constants already set: a
 * presentational component MAY hold its own `TextStyle` literals (raw `.sp` included) as long as
 * every *color* still comes from [FindlyTheme], and no §4.1 type-role or §4.1/§4.2 color-token
 * *name* is added, renamed, or bypassed. [CodeDisplayLetterSpacing]/[CodeDisplayFontFeatureSettings]
 * are exactly that: a component-local rendering detail, not a new contract vocabulary entry —
 * `FindlyTheme.typography.titleLarge` (already bold, already carrying its own -0.2sp tracking,
 * 003 §4.2) is the base every property is derived from via `.copy()`, never replaced.
 *
 * Used by [com.findly.android.ui.invites.CreateInviteScreen] for the create-invite success
 * view's large invite-code display; [code] is expected to already be in hyphenated display form
 * (`XXXX-XXXX`, 001-api-contract.md §1.4) — this component does no formatting of its own.
 */
private val CodeDisplayLetterSpacing = 4.sp

/** `"tnum"` (OpenType tabular-figures feature) — every digit renders at the same advance width,
 * so the code doesn't visually jitter/re-flow as different digits appear (HANDOFF.md's own
 * "tabular figures" ask, 010 §5.1). Not a color, not a §4.1 type role, not a spacing/corner/
 * elevation token — a raster/typesetting detail with no equivalent in the §4.1 vocabulary. */
private const val CodeDisplayFontFeatureSettings = "tnum"

@Composable
fun FindlyCodeDisplay(code: String, modifier: Modifier = Modifier) {
    Text(
        text = code,
        modifier = modifier,
        color = FindlyTheme.colors.onSurface,
        style = FindlyTheme.typography.titleLarge.copy(
            letterSpacing = CodeDisplayLetterSpacing,
            fontFeatureSettings = CodeDisplayFontFeatureSettings,
        ),
    )
}

File: mobile/ios/FindlyKit/Sources/FindlyKit/DesignSystem/Components/FindlyCodeDisplay.swift
import SwiftUI

/// specs/010-app-shell-and-screen-ux.md §5.1 item 1 ("titleLarge-class size, tabular figures,
/// letter-spaced, in hyphenated display form") — a small presentational component so the invite
/// code's styling lives in `DesignSystem/`, not hand-rolled at `CreateInviteScreen`'s call site.
/// Named `FindlyCodeDisplay` (review fix, I37 round 2), matching Android's independently-chosen
/// name for its own twin — every 010-batch component is `Findly*` on both platforms
/// (`FindlyNavDrawer`, `FindlyBottomSheet`, `FindlyDropdownField`); iOS's unprefixed components
/// (`StatusChip`, `MapMarkerBubble`, `EmptyStateView`) all predate this batch.
///
/// 004 §2.3's "zero styling outside DesignSystem/" rule scopes to `Screens/`, view models,
/// navigation, and networking — NOT to design-system components themselves, which are exactly
/// where presentational literals belong (`StatusChip`'s own `.tracking(0.3)` on its uppercase
/// label is the existing iOS precedent; Android's `FindlyMapMarkerBubble.kt` local `TextStyle`
/// literals are the cross-platform one). This component still reads `\.theme` only and derives
/// `titleLarge` rather than inventing a new typography token — `theme.typography.titleLarge.font`
/// supplies size/weight, `.monospacedDigit()` gives the tabular-figure alignment a code needs so
/// digits don't jitter width as they change, and an additional literal `.tracking(_:)` widens the
/// letter-spacing beyond what `titleLarge.tracking` alone provides (that token's own -0.2 is a
/// headline *tightening*, correct for a title, wrong for a code meant to read as visually
/// separated characters).
public struct FindlyCodeDisplay: View {
    @Environment(\.theme) private var theme
    private let displayForm: String

    /// [displayForm] is expected to already be the hyphenated `XXXX-XXXX` form (e.g.
    /// `CreateInviteViewModel.displayForm(for:)`) — this component renders whatever string it's
    /// given, with zero knowledge of the Crockford base32 code format itself.
    public init(_ displayForm: String) {
        self.displayForm = displayForm
    }

    public var body: some View {
        Text(displayForm)
            .font(theme.typography.titleLarge.font)
            .monospacedDigit()
            .tracking(2)
            .foregroundColor(theme.colors.onSurface)
            .accessibilityLabel("Invite code \(displayForm)")
    }
}

#Preview("FindlyCodeDisplay — light") {
    FindlyCodeDisplay("7F3K-9QRZ")
        .padding()
        .background(Theme.light.colors.surface)
        .environment(\.theme, .light)
}

#Preview("FindlyCodeDisplay — dark") {
    FindlyCodeDisplay("7F3K-9QRZ")
        .padding()
        .background(Theme.dark.colors.surface)
        .environment(\.theme, .dark)
        .preferredColorScheme(.dark)
}
